ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestination.java

```diff
@@ -248,22 +254,43 @@ private boolean updateBlobMetadata(BlobId blobId, String fieldName, Object value
     Objects.requireNonNull(fieldName, "Field name cannot be null");
 
     // We update the blob metadata value in two places:
-    // 1) the CosmosDB metadata collection
-    // 2) the blob storage entry metadata (to enable rebuilding the database)
-
+    // 1) the blob storage entry metadata (so GET's can be served entirely from ABS)
+    // 2) the CosmosDB metadata collection
     try {
-      if (!azureBlobDataAccessor.updateBlobMetadata(blobId, fieldName, value)) {
-        // TODO: what if this is a retry where ABS has been updated but Cosmos has not?
-        return false;
+      boolean updatedStorage, updatedCosmos = false;
+      AzureBlobDataAccessor.UpdateResponse updateResponse =
+          azureBlobDataAccessor.updateBlobMetadata(blobId, fieldName, value);
+      // Note: if blob does not exist will throw exception with NOT_FOUND status
+      Map<String, String> metadataMap = updateResponse.metadata;
+      updatedStorage = updateResponse.wasUpdated;
+
+      // Note: even if nothing changed in blob storage, still attempt to update Cosmos since this could be a retry
+      // of a request where ABS was updated but Cosmos update failed.
+      try {
+        ResourceResponse<Document> response = cosmosDataAccessor.updateMetadata(blobId, fieldName, value);
+        updatedCosmos = response != null;
+      } catch (DocumentClientException dex) {
+        if (dex.getStatusCode() == HttpConstants.StatusCodes.NOTFOUND) {
+          // blob exists in ABS but not Cosmos - inconsistent state
+          // Recover by inserting the updated map into cosmos
+          cosmosDataAccessor.upsertMetadata(CloudBlobMetadata.fromMap(metadataMap));
```

Comment:
I just want to check if there are any conditions where it may not be good to create a record. For example, what if compaction already cleaned up the record (or is that sequence of events impossible)?

Reply:
Good question.  It is possible for a blob to be purged but the Cosmos record not yet purged.  In this case, is it even possible for a lifecycle event to be passed here?  I would think the ReplicaThread would cut it off earlier, but not sure if the undelete feature changes that.  In that specific case, either expirationTime or deletionTime should be older than the retention period.